### ISSUE TITLE
Supporting CommonJS `require('../')` patterns where the directory is itself mapped, by supporting recursive package map

### ISSUE THREAD

**Issue by guybedford** [systemjs/systemjs#1164]
For the CommonJS configuration case with package configuration:

```javascript
{
  map: {
    './lib.js': './lib/index.js'
  }
}
```

Where `lib/dep.js` has `require('.')` which should resolve to `lib/index.js` the issue here is that that instead resolves to `lib/` as it is seen as a directory require in SystemJS.

We don't use the configuration:

```javascript
{
  map: {
    './lib/': './lib/index.js'
  }
}
```

here because that is circular map config, which means that `require('pkg/lib/index.js')` gets mapped to `pkg/lib/lib/index.js`.

The options for supporting this are thus:

1. Always have `System.normalizeSync('./x/', 'full/path/to/pkg/')` resolve to add a `.js` extension even if it has a trailing `/`. This unfortunately means normalizeSync can't be used to work out directory locations for things like asset loading, which was the use case in mind in specially not doing this here.
2. Allow recursive maps, but first detect if we're already inside the recursive map target folder and skip the recursive map in that case.

Perhaps (2) really is the best solution to be flexible. We already test for recursive maps as part of validation map configuration, so this should in theory only be a variation on that.

**Comment by guybedford** [comment]
Another option here might be to allow the trailing `/` in the map to indicate that `./lib/x` does not get mapped, and that only `./lib` and `./lib/` get mapped into `./lib/index.js`.

**Comment by guybedford** [comment]
The above is a little confusing as it conflicts with the meaning of the trailing `/` for paths.

I guess the point with recursive maps is that if `./x/y` (or `./x/y/`) maps to `./x/y/z` then we never do that map within the path `./x/y/*`, effectively making it a direct folder-only map anyway like we would get with that trailing `/` syntax, and it's a simple check to do.

Note `./x/y -> ./x` becomes an impossible map then, although I'm not sure that ever would have made sense.

**Comment by guybedford** [comment]
Released in 0.19.30.
